### PR TITLE
Add new "taskqlatency" script to observe taskq latencies via bpftrace

### PR DESCRIPTION
**Pull request by prakashsurya**
This series of commits allows us to get some more information about taskq latencies; i.e. it currently emits the latency spent queued on the taskq, and the latency spent executing the taskq entry's function.

I'm opening this review to solicit feedback about the correctness of the probes/script, the usability requirements for a script like this (e.g. potential command line flags, etc), better use of bpftrace's features (this is my first foray into using bpftrace), and/or anything else that might be relevant.

Example, showing latency spent queued on the taskq:
```
$ sudo timeout 15 ./contrib/bpftrace/taskqlatency.sh
Attaching 6 probes...
...
@queue_lat_us[dp_sync_taskq]:                                                  
[4K, 8K)               2 |@@@@@@                                              |
[8K, 16K)              5 |@@@@@@@@@@@@@@@@@                                   |
[16K, 32K)            11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@              |
[32K, 64K)             6 |@@@@@@@@@@@@@@@@@@@@                                |
[64K, 128K)           14 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@    |
[128K, 256K)           5 |@@@@@@@@@@@@@@@@@                                   |
[256K, 512K)          15 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[512K, 1M)            11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@              |
[1M, 2M)               8 |@@@@@@@@@@@@@@@@@@@@@@@@@@@                         |
[2M, 4M)              10 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                  |
[4M, 8M)               8 |@@@@@@@@@@@@@@@@@@@@@@@@@@@                         |
[8M, 16M)              2 |@@@@@@                                              |
                                                                               
@queue_lat_us[spl_dynamic_taskq]:                                              
[16K, 32K)             2 |@@@@@                                               |
[32K, 64K)             1 |@@                                                  |
[64K, 128K)            3 |@@@@@@@@                                            |
[128K, 256K)           4 |@@@@@@@@@@                                          |
[256K, 512K)           6 |@@@@@@@@@@@@@@@@                                    |
[512K, 1M)            13 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                 |
[1M, 2M)              11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                      |
[2M, 4M)              19 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4M, 8M)              17 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@      |
[8M, 16M)             12 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                    |
[16M, 32M)            12 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                    |
                                                                               
@queue_lat_us[z_rd_int]:                                                       
[16K, 32K)            13 |@@                                                  |
[32K, 64K)           245 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[64K, 128K)           24 |@@@@@                                               |
[128K, 256K)           6 |@                                                   |
[256K, 512K)          22 |@@@@                                                |
[512K, 1M)             3 |                                                    |
                                                                               
@queue_lat_us[z_wr_iss]:                                                       
[2K, 4K)               1 |                                                    |
[4K, 8K)              16 |@@@@                                                |
[8K, 16K)             19 |@@@@@                                               |
[16K, 32K)            28 |@@@@@@@@                                            |
[32K, 64K)            47 |@@@@@@@@@@@@@@                                      |
[64K, 128K)          137 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@          |
[128K, 256K)         168 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[256K, 512K)          70 |@@@@@@@@@@@@@@@@@@@@@                               |
[512K, 1M)            34 |@@@@@@@@@@                                          |
                                                                               
@queue_lat_us[z_wr_int]:                                                       
[4K, 8K)               8 |@                                                   |
[8K, 16K)            134 |@@@@@@@@@@@@@@@@@@@@                                |
[16K, 32K)           336 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[32K, 64K)           269 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@           |
[64K, 128K)          157 |@@@@@@@@@@@@@@@@@@@@@@@@                            |
[128K, 256K)          93 |@@@@@@@@@@@@@@                                      |
[256K, 512K)          41 |@@@@@@                                              |
[512K, 1M)             7 |@                                                   |
[1M, 2M)               3 |                                                    |
...
```

Example, showing latency spent executing the taskq entry's function:
```
$ sudo timeout 15 ./contrib/bpftrace/taskqlatency.sh
Attaching 6 probes...
...
@exec_lat_us[dp_sync_taskq, userquota_updates_task]:                           
[2K, 4K)              10 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4K, 8K)               3 |@@@@@@@@@@@@@@@                                     |
[8K, 16K)              7 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                |
[16K, 32K)             5 |@@@@@@@@@@@@@@@@@@@@@@@@@@                          |
[32K, 64K)             2 |@@@@@@@@@@                                          |
                                                                               
@exec_lat_us[metaslab_group_taskq, metaslab_preload]:                          
[1K, 2K)              27 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[2K, 4K)               3 |@@@@@                                               |
[4K, 8K)               3 |@@@@@                                               |
                                                                               
@exec_lat_us[dp_sync_taskq, sync_dnodes_task]:                                 
[2K, 4K)               3 |@@@@@@@@@@@                                         |
[4K, 8K)               8 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                       |
[8K, 16K)              4 |@@@@@@@@@@@@@@                                      |
[16K, 32K)             0 |                                                    |
[32K, 64K)             2 |@@@@@@@                                             |
[64K, 128K)            7 |@@@@@@@@@@@@@@@@@@@@@@@@@@                          |
[128K, 256K)          11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@            |
[256K, 512K)          11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@            |
[512K, 1M)            14 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[1M, 2M)               2 |@@@@@@@                                             |
[2M, 4M)               5 |@@@@@@@@@@@@@@@@@@                                  |
[4M, 8M)               3 |@@@@@@@@@@@                                         |
                                                                               
@exec_lat_us[spl_dynamic_taskq, taskq_thread_spawn_task]:                      
[8K, 16K)             10 |@@@@@@@@@@@@@@@@@@@@@@@@@@@                         |
[16K, 32K)            10 |@@@@@@@@@@@@@@@@@@@@@@@@@@@                         |
[32K, 64K)             4 |@@@@@@@@@@                                          |
[64K, 128K)            3 |@@@@@@@@                                            |
[128K, 256K)           3 |@@@@@@@@                                            |
[256K, 512K)          11 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                      |
[512K, 1M)            16 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@         |
[1M, 2M)              13 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@                 |
[2M, 4M)              19 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[4M, 8M)              14 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@              |
                                                                               
@exec_lat_us[z_rd_int, zio_execute]:                                           
[8K, 16K)              5 |@                                                   |
[16K, 32K)            34 |@@@@@@@                                             |
[32K, 64K)           231 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@|
[64K, 128K)           14 |@@@                                                 |
[128K, 256K)           5 |@                                                   |
[256K, 512K)          25 |@@@@@                                               |
...
```

### CODE REVIEW

File: contrib/bpftrace/taskqlatency.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        @tq_name[tid] = str($tqent->tqent_taskq->tq_name);
+
+        @start[@tq_name[tid], @tqent_id[tid]] = nsecs;
+}
+
+kprobe:trace_zfs_taskq_ent__finish
+/ @birth[@tq_name[tid], @tqent_id[tid]] /
+{
+        @queue_lat_us[@tq_name[tid]] =
+                hist((nsecs - @birth[@tq_name[tid], @tqent_id[tid]]));
```
Comment (ahrens) [inline contrib/bpftrace/taskqlatency.sh]: The aggregations are named `*_us`, implying that they are microseconds, but they are actually in nanoseconds.

Comment (ahrens) [inline contrib/bpftrace/taskqlatency.sh]: I would think of the time "queued" as the time that the entry is on the queue, i.e. the time between when taskq_dispatch() is called and when the callback is called (i.e. begins executing).  But this looks like it's defining "queued" as the time between dispatch and when the callback *finishes* executing.

File: contrib/bpftrace/taskqlatency.sh [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+}
+EOF
+
+KVER=$(uname -r)
+bpftrace \
+        --include "/usr/src/zfs-$KVER/zfs_config.h" \
+        -I "/usr/src/zfs-$KVER/include" \
+        -I "/usr/src/zfs-$KVER/include/os/linux/kernel" \
+        -I "/usr/src/zfs-$KVER/include/os/linux/spl" \
+        -I "/usr/src/zfs-$KVER/include/os/linux/zfs" \
```
Comment (ahrens) [inline contrib/bpftrace/taskqlatency.sh]: FYI: once I integrate https://github.com/delphix/zfs/pull/92 these won't be needed

File: contrib/bpftrace/taskqlatency.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        clear(@tq_name);
+        clear(@tqent_id);
+}
+EOF
+
+KVER=$(uname -r)
+bpftrace \
+        --include "/usr/src/zfs-$KVER/zfs_config.h" \
+        -I "/usr/src/zfs-$KVER/include" \
```
Comment (ahrens) [inline contrib/bpftrace/taskqlatency.sh]: It's kind of annoying to have this bash wrapper that doesn't really do that much.  I think we should consider making a standalone script that would invoke bpftrace with the right arguments to use with ZFS (possible names `zbpf` or `bpftrace_zfs`).  Then your script could use `#!/usr/bin/zbpf`

File: lib/libzpool/Makefile.am [hunk 4]
```diff
@@ -106,7 +106,6 @@ KERNEL_C = \
 	space_map.c \
 	space_reftree.c \
 	txg.c \
-	trace.c \
```
Comment (ahrens) [inline lib/libzpool/Makefile.am]: Is this related to your other changes?  Or was it never really required?  Maybe the comment in trace.c needs to be updated?

File: contrib/bpftrace/taskqlatency.sh (hunk 3 above)
Comment (prakashsurya) [reply contrib/bpftrace/taskqlatency.sh]: I agree; but I'm not sold on the addition of some new script that's a wrapper around `bpftrace`, but adds the `-I` flag.. I'd like to keep what I have here for now, until we have a solid plan for how to address this moving forward; then we can retrofit this script to use whatever mechanism we decided long term.

For example, I wonder if it'd be better to augment `bpftrace` itself to look for headers in the ZFS directory when the directory exists, but I don't know how feasible that would be without having more time to investigate that solution.

I'd like to avoid rushing to whatever solution is easiest to implement (now), without first trying to think about the most correct solution, and pursuing that first.

File: lib/libzpool/Makefile.am (hunk 4 above)
Comment (prakashsurya) [reply lib/libzpool/Makefile.am]: I had to remove it, because this file no longer exists; `trace.c` was renamed to `spl-trace.c` and moved into the SPL directory.

**Comment by prakashsurya** [comment]
@ahrens I added a new commit to address most of your feedback; let me know what you think when you have a chance to give it a look.

File: contrib/bpftrace/taskqlatency.sh (hunk 3 above)
Comment (prakashsurya) [reply contrib/bpftrace/taskqlatency.sh]: I haven't tested this, but I've opened https://github.com/delphix/bpftrace/pull/4 .. Let me know what you think, and if I should spend time to test and verify such an approach would work.

Comment (prakashsurya) [reply contrib/bpftrace/taskqlatency.sh]: This probably isn't the correct place to have this discussion, but it made me remember and realize that we only keep the "latest" set of headers installed on the system at any given time. So, for example, if we install a new set of headers but don't reboot to start running the modules that correspond to those headers, I think this approach may not work (I don't know if it'll work by "accident", silently return the wrong information, or explicitly fail).

File: lib/libzpool/Makefile.am (hunk 4 above)
Comment (prakashsurya) [reply lib/libzpool/Makefile.am]: This must be a quirk of the github review UI..

If you look at [this commit](https://github.com/delphix/zfs/pull/95/commits/8b8507285371e8828de30e6259074b353d2850e4) it shows the file has been "renamed without changes".

But if you look at all the changes for the PR at once, [here](https://github.com/delphix/zfs/pull/95/files), it doesn't show any rename; instead makes it appear as I'm modifying the `module/os/linux/spl/spl-taskq.c` file, and that file was already at that location (i.e. I didn't do any rename).

The more I use the github UI for reviewing changes, the more I distrust it.

File: contrib/bpftrace/taskqlatency.sh (hunk 3 above)
Comment (ahrens) [reply contrib/bpftrace/taskqlatency.sh]: The upgrade-but-not-rebooted case is interesting, thanks for bringing that up.

If the kernel was also updated, then I think everything would "just work" since `uname -r` / `utsname.release` and thus the path to the zfs includes would be different for the newer version, and we'd keep using the older include files.

But if the kernel version was exactly the same, then the new ZFS headers would be copied over the old ones, right?  In which case we'd be using the newer headers with the older (still running) kernel module.

Would it be possible to include the Delphix release in `uname -r` somehow?  Or should we change the path to include the ZFS module "version" somehow?

File: lib/libzpool/Makefile.am (hunk 4 above)
Comment (ahrens) [reply lib/libzpool/Makefile.am]: Ah, I missed that.  Are you planning to submit this change upstream to ZoL?  It would be nice to not have too much diffs in the build infrastructure.

Comment (prakashsurya) [reply lib/libzpool/Makefile.am]: Yes; if you're OK with this change, I can close this PR and open it upstream. I wanted to get some feedback with our team first, before pushing it upstream.

**Comment by prakashsurya** [comment]
Closing in favor of the PR I opened upstream: https://github.com/zfsonlinux/zfs/pull/9525

File: contrib/bpftrace/taskqlatency.sh (hunk 3 above)
Comment (prakashsurya) [reply contrib/bpftrace/taskqlatency.sh]: I had opened a bug about this a while back, here's a link to it: https://jira.delphix.com/browse/DLPX-65703
